modules/contact/src/linesearches/PetscPingPongLineSearch.C

```diff
@@ -0,0 +1,173 @@
+}
+
+void
+PetscPingPongLineSearch::initialSetup()
+{
+  _displaced_problem = _fe_problem.getDisplacedProblem().get();
+  if (!_displaced_problem)
+    mooseError("This line search only makes sense in a displaced context");
+}
+
+void
+PetscPingPongLineSearch::lineSearch()
+{
+  PetscBool changed_y = PETSC_FALSE;
+  PetscErrorCode ierr;
+  Vec X, F, Y, W, G;
+  SNESLineSearch line_search;
+  PetscReal fnorm, xnorm, ynorm;
+  PetscBool domainerror;
+  SNES snes = _solver->snes();
+
+  ierr = SNESGetLineSearch(snes, &line_search);
+  LIBMESH_CHKERR(ierr);
+  ierr = SNESLineSearchGetVecs(line_search, &X, &F, &Y, &W, &G);
+  LIBMESH_CHKERR(ierr);
+  ierr = SNESLineSearchGetNorms(line_search, &xnorm, &fnorm, &ynorm);
+  LIBMESH_CHKERR(ierr);
+  ierr = SNESLineSearchSetReason(line_search, SNES_LINESEARCH_SUCCEEDED);
+  LIBMESH_CHKERR(ierr);
+
+  ierr = SNESLineSearchPreCheck(line_search, X, Y, &changed_y);
+  LIBMESH_CHKERR(ierr);
+
+  // apply the full newton step
+  ierr = VecWAXPY(W, -1., Y, X);
+  LIBMESH_CHKERR(ierr);
+
+  {
+    PetscVector<Number> solution(W, this->comm());
+
+    _nl.setSolution(solution);
+
+    // Displace the mesh and update the displaced geometric search objects
+    _displaced_problem->updateMesh();
+  }
+
+  // A reference to the displaced MeshBase object
+  const auto & mesh = _displaced_problem->mesh().getMesh();
+
+  const auto & pen_locs = _displaced_problem->geomSearchData().getPenetrationLocators();
+
+  // Keep track of the slave nodes that we push back onto the master face. We'll eventually check to
+  // make sure that we didn't have a corner node at the intersection of two slave faces that we
+  // tried to displace twice. As this stands now this won't cover the case wherethe intersection
+  // happens only across processes
+  std::set<dof_id_type> nodes_displaced;
+
+  std::vector<unsigned int> disp_nums;
+
+  // Generate the displaced variable numbers
+  for (const auto & disp_name : _displaced_problem->getDisplacementVarNames())
+    disp_nums.push_back(_nl.system().variable_number(disp_name));
+
+  for (const auto & pen_loc : pen_locs)
+    for (const auto & pinfo_pair : pen_loc.second->_penetration_info)
+    {
+      auto node_id = pinfo_pair.first;
+      auto pen_info = pinfo_pair.second;
+
+      // We have penetration
+      if (pen_info->_distance > 0)
+      {
+// Avoid warning in optimized modes about unused variables
+#ifndef NDEBUG
+        // Make sure we haven't done this node before
+        auto pair = nodes_displaced.insert(node_id);
+        mooseAssert(pair.second, "Node id " << node_id << " has already been displaced");
+#endif
+
+        const auto & node = mesh.node_ref(node_id);
+
+        // If this is not a local node, we will let displacement happen on another process
+        if (node.processor_id() != this->processor_id())
+          continue;
+
+        // The vector that we need to displace by
+        auto required_solution_change = pen_info->_distance * pen_info->_normal;
```

Comment:
How do you figure out it is the solution update? Should we change the function directly so that the update is right automatically?

Reply:
I might have to call and talk to you about this one. Too much to type